test/gtest/ucp/test_ucp_am.cc

```diff
@@ -0,0 +1,290 @@
+    std::vector<char> cmp(length, (char)length);
+    std::vector<char> databuf(length, 'r');
+
+    memcpy(&databuf[0], data, length);
+
+    EXPECT_EQ(cmp, databuf);
+    if (me->release) {
+        me->for_release[me->recv_ams] = data;
+        status                        = UCS_INPROGRESS;
+    } else {
+        status = UCS_OK;
+    }
+    me->recv_ams++;
+    return status;
+}
+
+class test_ucp_am : public test_ucp_am_base
+{
+    public:
+        ucp_ep_params_t get_ep_params() {
+            ucp_ep_params_t params = test_ucp_am_base::get_ep_params();
+            params.field_mask     |= UCP_EP_PARAM_FIELD_FLAGS;
+            params.flags          |= UCP_EP_PARAMS_FLAGS_NO_LOOPBACK;
+            return params;
+        }
+        virtual void init(){
+            ucp_test::init();
+            sender().connect(&receiver(), get_ep_params());
+            receiver().connect(&sender(), get_ep_params());
+        }
+
+    protected:
+        void do_set_am_handler_realloc_test();
+        void do_send_process_data_test(int test_release, uint16_t am_id,
+                                       int send_reply);
+        void do_send_process_data_iov_test();
+        void set_handlers(uint16_t am_id);
+        void set_reply_handlers(uint16_t am_id);
+};
+
+void test_ucp_am::set_reply_handlers(uint16_t am_id)
+{
+    ucp_worker_set_am_handler(sender().worker(), am_id,
+                              ucp_process_reply_cb, this,
+                              UCP_AM_FLAG_WHOLE_MSG);
+    ucp_worker_set_am_handler(receiver().worker(), am_id,
+                              ucp_process_reply_cb, this,
+                              UCP_AM_FLAG_WHOLE_MSG);
+}
+
+void test_ucp_am::set_handlers(uint16_t am_id)
+{
+    ucp_worker_set_am_handler(sender().worker(), am_id,
+                              ucp_process_am_cb, this,
+                              UCP_AM_FLAG_WHOLE_MSG);
+    ucp_worker_set_am_handler(receiver().worker(), am_id,
+                              ucp_process_am_cb, this,
+                              UCP_AM_FLAG_WHOLE_MSG);
+}
+
+void test_ucp_am::do_send_process_data_test(int test_release, uint16_t am_id,
+                                            int send_reply)
+{
+    size_t buf_size = pow(2, NUM_MESSAGES - 2); /* minus 2 because 0 and 1*/
+    ucs_status_ptr_t sstatus = NULL;
+    std::vector<char> buf(buf_size);
+
+    recv_ams      = 0;
+    sent_ams      = 0;
+    replies       = 0;
+    this->release = test_release;
+
+    for (size_t i = 0; i < buf_size + 1; i *= 2) {
+        for (size_t j = 0; j < i; j++) {
+            buf[j] = i;
+        }
+        reply   = NULL;
+        sstatus = ucp_am_send_nb(receiver().ep(), am_id,
+                                 buf.data(), 1, ucp_dt_make_contig(i),
+                                 test_ucp_am_base::ucp_send_am_cb,
+                                 send_reply);
+
+        EXPECT_FALSE(UCS_PTR_IS_ERR(sstatus));
+        wait(sstatus);
+
+        if(i == 0) {
```

Comment:
why changing the iteration variable, i, inside the for loop??
seems like bad practice

Reply:
I start with a 0 byte message and the loop doubles the size every iteration, but 0 * 2 is still 0, so I have to increase it. I can start at 1 to remove those, or move the 0 byte message.

Reply:
how about: instead of 
`i*=2`
do
`i = i ? (i * 2) : 1`

Reply:
That is definitely better. Thank you.